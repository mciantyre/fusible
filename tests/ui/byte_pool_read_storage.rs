// SPDX-License-Identifier: MPL-2.0
// SPDX-FileCopyrightText: Copyright 2024 Ian McIntyre

use fusible::byte_pool::BytePool;

fn use_storage(_: &[u8]) {}

/// Once you create a pool with storage, the pool exclusively
/// borrows that storage. You can't read from the storage.
fn main() {
    let mut storage = [0u8; 1024];
    let pool = core::pin::pin!(BytePool::context());
    BytePool::create(pool.into_ref(), &mut storage, &Default::default()).unwrap();
    use_storage(&storage);
}
