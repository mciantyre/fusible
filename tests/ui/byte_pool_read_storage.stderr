error[E0503]: cannot use `storage` because it was mutably borrowed
  --> tests/ui/byte_pool_read_storage.rs:12:13
   |
10 |     let pool = core::pin::pin!(BytePool::context());
   |                ------------------------------------ a temporary with access to the borrow is created here ...
11 |     BytePool::create(pool.into_ref(), &mut storage, &Default::default()).unwrap();
   |                                       ------------ `storage` is borrowed here
12 |     let _ = storage[0];
   |             ^^^^^^^^^^ use of borrowed `storage`
13 | }
   | - ... and the borrow might be used here, when that temporary is dropped and runs the `Drop` code for type `BytePoolContext`
